Calling getUserMedia with echo cancellation disabled causes side effects for existing audio tracks acquired via getUserMedia

youennf opened the pull request:
#### 9c7c6b9984936126323da202f62288548cc28fb0
<pre>
Calling getUserMedia with echo cancellation disabled causes side effects for existing audio tracks acquired via getUserMedia
<a href="https://rdar.apple.com/151143554">rdar://151143554</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292861">https://bugs.webkit.org/show_bug.cgi?id=292861</a>

Reviewed by Jean-Yves Avenard.

On macOS, we can run one VPIO unit and several HAL units in parallel.
This allows to use at the same time a microphone without echo cancellation (using HAL units) and with echo cancellation (using VPIO).

To do so, CoreAudioCaptureSource has its own CoreAudioCaptureUnit whenever echo cancellation is off.
CoreAudioCaptureSource uses CoreAudioCaptureUnit::defaultSingleton whenever echo cancellation is on.

We update BaseAudioCaptureUnit to take a boolean telling whether it supports echo cancellation.
This is used in BaseAudioCaptureUnit::setEnableEchoCancellation to validate we are not creating more than one VPIO unit.

CoreAudioCaptureSource, just before initializing its audio unit, will select its CoreAudioCaptureUnit based on echo cancellation.
The same principle is also done in CoreAudioCaptureSource::settingsDidChange, which happens when the web page calls MediaStreamTrack.applyConstraints.
The code path used for applyConstraints reuse the initialization code path, which requires to reset m_shouldInitializeAudioUnit to true.

It is possible to use applyConstraints to switch between HAL and VPIO units but all track clones are sharing the same CoreAudioCaptureUnit.
To get independent tracks, the web page will have to call getUserMedia twice.
A future patch should fix this limitation.

This is specific to macOS, no change to iOS behavior.

Now that we are creating potentially more than one CoreAudioCaptureUnit, we need to mock all of them.
We update our mock code so that we can iterate on existing CoreAudioCaptureUnit and update newly created.

Covered by added test.

Canonical link: <a href="https://commits.webkit.org/302582@main">https://commits.webkit.org/302582@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bd6ad5cd907b99134184b1aa5bce219e49e7d85

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129583 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1840 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136969 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/81022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9284ed7d-840a-4a3a-a989-af193813ecae) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1772 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1716 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/81022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/efb4a1a8-31c5-4db4-979a-4ff5d1f1986b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1374 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116069 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79373 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c56decac-c392-46b2-b112-5bdcb47dbebd) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34200 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80244 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109766 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139449 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1630 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1559 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107229 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1672 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112410 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107074 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27258 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1331 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/30922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54349 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1701 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65064 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1521 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1555 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1623 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureUnit.cpp:
```diff
@@ -230,13 +235,27 @@ void CoreAudioCaptureUnit::forEach(NOESCAPE Function<void(CoreAudioCaptureUnit&)
 
-CoreAudioCaptureUnit::CoreAudioCaptureUnit()
-    : m_sampleRateCapabilities(8000, 96000)
+static Function<void(CoreAudioCaptureUnit&)>& coreAudioCaptureNewUnitCallback()
```
```suggestion
static Function<void(CoreAudioCaptureUnit&)>& coreAudioCaptureNewUnitCallbackSingleton()
```

jyavenard replied:
should we assert that we are on a set thread? (main thread?) this isn't thread-safe

youennf replied:
Yes, this is main thread only, I'll add an ASSERT since this source file is doing that already.

youennf committed:
Calling getUserMedia with echo cancellation disabled causes side effects for existing audio tracks acquired via getUserMedia
rdar://151143554
https://bugs.webkit.org/show_bug.cgi?id=292861

Reviewed by Jean-Yves Avenard.

On macOS, we can run one VPIO unit and several HAL units in parallel.
This allows to use at the same time a microphone without echo cancellation (using HAL units) and with echo cancellation (using VPIO).

To do so, CoreAudioCaptureSource has its own CoreAudioCaptureUnit whenever echo cancellation is off.
CoreAudioCaptureSource uses CoreAudioCaptureUnit::defaultSingleton whenever echo cancellation is on.

We update BaseAudioCaptureUnit to take a boolean telling whether it supports echo cancellation.
This is used in BaseAudioCaptureUnit::setEnableEchoCancellation to validate we are not creating more than one VPIO unit.

CoreAudioCaptureSource, just before initializing its audio unit, will select its CoreAudioCaptureUnit based on echo cancellation.
The same principle is also done in CoreAudioCaptureSource::settingsDidChange, which happens when the web page calls MediaStreamTrack.applyConstraints.
The code path used for applyConstraints reuse the initialization code path, which requires to reset m_shouldInitializeAudioUnit to true.

It is possible to use applyConstraints to switch between HAL and VPIO units but all track clones are sharing the same CoreAudioCaptureUnit.
To get independent tracks, the web page will have to call getUserMedia twice.
A future patch should fix this limitation.

This is specific to macOS, no change to iOS behavior.

Now that we are creating potentially more than one CoreAudioCaptureUnit, we need to mock all of them.
We update our mock code so that we can iterate on existing CoreAudioCaptureUnit and update newly created.

Covered by added test.

Canonical link: https://commits.webkit.org/302582@main
